File: apps/api/app/schemas/monitor.py
"""Monitor schemas."""
from typing import Optional
from pydantic import BaseModel, Field


class MonitorRuleCreate(BaseModel):
    type: str
    name: str
    description: str = ""
    isEnabled: bool = True
    threshold: dict = {}
    targetAddresses: list = []
    targetFunctions: list = []
    targetEvents: list = []
    notificationChannels: list = ["in_app"]


class MonitorCreateRequest(BaseModel):
    contractId: str
    name: str
    description: Optional[str] = None
    chainId: str = "stellar-testnet"
    contractAddress: str
    checkInterval: int = 300  # seconds
    rules: list[MonitorRuleCreate] = []


class MonitorResponse(BaseModel):
    id: str
    userId: str
    contractId: str
    name: str
    description: Optional[str] = None
    chainId: str
    contractAddress: str
    isActive: bool = True
    checkInterval: int
    rules: list = []
    lastCheckedAt: Optional[str] = None
    totalAlerts: int = 0
    createdAt: str
    updatedAt: str


class MonitorEventResponse(BaseModel):
    id: str
    monitorId: str
    txHash: str
    blockNumber: int
    timestamp: str
    ruleType: str
    description: str
    data: dict = {}

File: packages/shared/src/types/monitor.ts
export interface Monitor {
  id: string;
  userId: string;
  contractId: string;
  name: string;
  description: string | null;
  chainId: string;
  contractAddress: string;
  isActive: boolean;
  checkInterval: number;
  rules: MonitorRule[];
  lastCheckedAt: string | null;
  totalAlerts: number;
  createdAt: string;
  updatedAt: string;
}

export enum MonitorRuleType {
  OWNERSHIP_TRANSFER = "ownership_transfer",
  PAUSE_UNPAUSE = "pause_unpause",
  PROXY_UPGRADE = "proxy_upgrade",
  TOKEN_MINT = "token_mint",
  LARGE_WITHDRAWAL = "large_withdrawal",
  SUSPICIOUS_TX_SPIKE = "suspicious_tx_spike",
  FUNCTION_CALL = "function_call",
  BALANCE_CHANGE = "balance_change",
  EVENT_EMISSION = "event_emission",
  CUSTOM = "custom",
}

export interface MonitorRule {
  id: string;
  type: MonitorRuleType;
  name: string;
  description: string;
  isEnabled: boolean;
  threshold: MonitorThreshold;
  targetAddresses: string[];
  targetFunctions: string[];
  targetEvents: string[];
  notificationChannels: string[];
  metadata: Record<string, unknown>;
}

export interface MonitorThreshold {
  minAmount: string | null;
  maxAmount: string | null;
  minFrequency: number | null;
  timeWindow: number;
  alertCooldown: number;
}

export interface MonitorEvent {
  id: string;
  monitorId: string;
  txHash: string;
  blockNumber: number;
  timestamp: string;
  ruleType: MonitorRuleType;
  description: string;
  data: Record<string, unknown>;
}
